File: data-structures/linked-list/circular-doubly-linked-list/CircularDoublyLinkedList.java
class CircularDoublyLinkedList {
  Node head, tail;

  class Node {
    int data;
    Node next;
    Node prev;

    Node(int data) {
      this.data = data;
      this.next = this.prev = null;
    }
  }

  public CircularDoublyLinkedList() {
    tail = null;
    head = null;
  }

  void insertFirst(int data) {
    Node newNode = new Node(data);

    if (head == null) {
      head = tail = newNode;
      head.next = head;
      head.prev = head;
    } else {
      newNode.next = head;
      newNode.prev = tail;
      head.prev = newNode;
      tail.next = newNode;
      head = newNode;
    }
  }

  void insertLast(int data) {
    Node newNode = new Node(data);

    if (head == null) {
      head = tail = newNode;
      head.next = head;
      tail.prev = head;
    } else {
      tail.next = newNode;
      newNode.prev = tail;
      newNode.next = head;
      head.prev = newNode;
      tail = newNode;
    }
  }

  public void deleteFirst() {
    if (head == null) {
      System.out.println("List is empty");
      return;
    }

    if (head == tail) {
      head = tail = null;
    } else {
      head = head.next;
      head.prev = tail;
      tail.next = head;
    }
  }

  public void deleteLast() {
    if (tail == null) {
      System.out.println("List is empty");
      return;
    }

    if (head == tail) {
      head = tail = null;
    } else {
      tail = tail.prev;
      tail.next = head;
      head.prev = tail;
    }
  }

  public void displayForward() {
    if (head == null) {
      System.out.println("List is empty");
      return;
    }

    Node currNode = head;
    System.out.print("Forward: ");

    do {
      System.out.print(currNode.data + " ");
      currNode = currNode.next;
    } while (currNode != head);

    System.out.println();
  }

  public void displayBackward() {
    if (tail == null) {
      System.out.println("List is empty");
      return;
    }

    Node currNode = tail;
    System.out.print("Backward: ");

    do {
      System.out.print(currNode.data + " ");
      currNode = currNode.prev;
    } while (currNode != tail);

    System.out.println();
  }
}

File: data-structures/linear/linked-list/circular-doubly-linked-list/CircularDoublyLinkedList.cpp
#include <iostream>
using namespace std;

class Node {
public:
  int data;
  Node *next;
  Node *prev;

  Node(int value) {
    data = value;
    next = prev = nullptr;
  }
};

class CircularDoublyLinkedList {
  Node *head;
  Node *tail;

public:
  CircularDoublyLinkedList() { head = tail = nullptr; }

  void insertAtHead(int value) {
    Node *newNode = new Node(value);

    if (head == nullptr) {
      head = tail = newNode;
      head->next = head->prev = head;
    } else {
      newNode->next = head;
      newNode->prev = tail;
      head->prev = newNode;
      tail->next = newNode;
      head = newNode;
    }
  }

  void insertAtTail(int value) {
    Node *newNode = new Node(value);

    if (head == nullptr) {
      head = tail = newNode;
      head->next = head->prev = head;
    } else {
      newNode->prev = tail;
      newNode->next = head;
      tail->next = newNode;
      head->prev = newNode;
      tail = newNode;
    }
  }

  void deleteAtHead() {
    if (head == nullptr) {
      return;
    } else if (head == tail) {
      delete head;
      head = tail = nullptr;
    } else {
      Node *temp = head;

      head = head->next;
      head->prev = tail;
      tail->next = head;
      temp->next = temp->prev = nullptr;
      delete temp;
    }
  }

  void deleteAtTail() {
    if (head == nullptr) {
      return;
    } else if (head == tail) {
      delete head;
      head = tail = nullptr;
    } else {
      Node *temp = tail;

      tail = tail->prev;
      tail->next = head;
      head->prev = tail;
      temp->next = temp->prev = nullptr;
      delete temp;
    }
  }

  void print() {
    if (head == nullptr) {
      return;
    }

    Node *temp = head;

    do {
      cout << temp->data << " <-> ";
      temp = temp->next;
    } while (temp != head);

    cout << temp->data << "\n";
  }
};
